# Phase 5: Evaluation

**CRISP-DM Phase Description:**  
At this stage, the model(s) appear to have high quality from a technical standpoint. Before proceeding to final deployment, it is important to evaluate the model more thoroughly and review the steps that were executed to construct it, to be certain the model properly achieves the business objectives. A key objective is to determine if there is some important business issue that has not been sufficiently considered.

---

In [9]:
# Standard imports
import pandas as pd
import numpy as np

---
### Task 1: Evaluate Results

Assess how well the model(s) meet the **business success criteria** originally defined in **Phase 1 (Business Understanding)**. This is distinct from the technical assessment in Phase 4 — here you ask: *"Does this model actually solve the business problem?"*

Key activities:
- **Map Technical Metrics to Business Criteria:** Compare the model's performance metrics (from Phase 4) against the business success criteria and data mining success criteria defined in Phase 1.
- **Business Impact Assessment:** Translate technical performance into business-relevant terms (e.g., "The model's 92% recall means we would catch 92 out of every 100 fraudulent transactions").
- **Gap Analysis:** Identify any gaps between expected and actual performance.
- **Approved Models:** List the model(s) that pass the evaluation and the ones that do not.

**Instructions:** Reference the `business_objectives` and `data_mining_goals` dictionaries from Phase 1. Compare each success criterion against the model's actual performance.

In [10]:
# TODO: Evaluate model results against business success criteria.
# Reference your Phase 1 definitions and Phase 4 results.

evaluation = {
    "business_criteria": [
        {
            "criterion": "Achieve at least 80% classification accuracy",
            "target": ">= 0.80 accuracy",
            "achieved": "0.3536 accuracy using Random Forest",
            "met": False
        },
        {
            "criterion": "Identify key features influencing crime types",
            "target": "Model should learn meaningful patterns from time and location features",
            "achieved": "Partially achieved through the use of spatial and temporal features such as district, beat, hour, day, and month",
            "met": True
        },
        {
            "criterion": "Provide insights into crime patterns across time and location",
            "target": "Generate interpretable findings to support analysis",
            "achieved": "Achieved through EDA and model evaluation, which revealed temporal and spatial crime patterns",
            "met": True
        }
    ],
    "technical_criteria": [
        {
            "metric": "Accuracy",
            "target": "As high as possible",
            "achieved": 0.353642,
            "met": True
        },
        {
            "metric": "Weighted Precision",
            "target": "As high as possible",
            "achieved": 0.300473,
            "met": True
        },
        {
            "metric": "Weighted Recall",
            "target": "As high as possible",
            "achieved": 0.353642,
            "met": True
        },
        {
            "metric": "Weighted F1-Score",
            "target": "Primary comparison metric due to class imbalance",
            "achieved": 0.313837,
            "met": True
        },
        {
            "metric": "Cross-Validation Mean F1",
            "target": "Stable across folds",
            "achieved": 0.3006,
            "met": True
        }
    ],
    "overall_assessment": "The model partially meets the business objectives. It provides useful insights into crime patterns and achieves stable performance, but the predictive accuracy is not high enough to fully satisfy the desired success criterion of 80% accuracy."
}

In [11]:
# Display the evaluation summary
print("=" * 60)
print("EVALUATION AGAINST BUSINESS SUCCESS CRITERIA")
print("=" * 60)

print("\n--- Business Criteria ---")
for c in evaluation["business_criteria"]:
    status = "PASS" if c["met"] else "FAIL"
    print(f"[{status}] {c['criterion']}")
    print(f"Target  : {c['target']}")
    print(f"Achieved: {c['achieved']}\n")

print("\n--- Technical Criteria ---")
for c in evaluation["technical_criteria"]:
    status = "PASS" if c["met"] else "FAIL"
    print(f"[{status}] {c['metric']}")
    print(f"Target  : {c['target']}")
    print(f"Achieved: {c['achieved']}\n")

print("\nOverall Assessment:")
print(evaluation["overall_assessment"])

EVALUATION AGAINST BUSINESS SUCCESS CRITERIA

--- Business Criteria ---
[FAIL] Achieve at least 80% classification accuracy
Target  : >= 0.80 accuracy
Achieved: 0.3536 accuracy using Random Forest

[PASS] Identify key features influencing crime types
Target  : Model should learn meaningful patterns from time and location features
Achieved: Partially achieved through the use of spatial and temporal features such as district, beat, hour, day, and month

[PASS] Provide insights into crime patterns across time and location
Target  : Generate interpretable findings to support analysis
Achieved: Achieved through EDA and model evaluation, which revealed temporal and spatial crime patterns


--- Technical Criteria ---
[PASS] Accuracy
Target  : As high as possible
Achieved: 0.353642

[PASS] Weighted Precision
Target  : As high as possible
Achieved: 0.300473

[PASS] Weighted Recall
Target  : As high as possible
Achieved: 0.353642

[PASS] Weighted F1-Score
Target  : Primary comparison metric due 

The Random Forest model achieved the best performance with an accuracy of approximately 35.36% and a weighted F1-score of 0.3138. Although this does not meet the original aspiration of 80% accuracy, the model still provides useful analytical value by capturing meaningful crime patterns from spatial and temporal features. Cross-validation results also show stable performance, which indicates that the model generalises reasonably well.

---
### Task 2: Review Process

Conduct a thorough review of the entire CRISP-DM process executed so far. The goal is to ensure quality assurance — that no important task or factor was overlooked. Consider:

- **Completeness Check:** Were all relevant data sources considered? Were the data preparation steps sufficient?
- **Methodology Review:** Were appropriate techniques used at each stage? Were there alternatives worth exploring?
- **Data Leakage Check:** Did any information from the test set leak into the training process?
- **Bias and Fairness:** Could the model introduce or amplify any biases? Is the model fair across different subgroups?
- **Reproducibility:** Can the entire pipeline be reproduced from scratch with consistent results?

**Instructions:** Complete the process review checklist below and document any concerns.

In [12]:
# TODO: Review the end-to-end process.

process_review = {
    "checklist": [
        {
            "item": "All relevant data sources were considered",
            "status": "Partial",
            "notes": "The project used a single crime dataset. Additional external data such as socio-economic or weather data could further improve performance."
        },
        {
            "item": "Data preparation was thorough and documented",
            "status": "Yes",
            "notes": "Missing values were handled, duplicates removed, outliers filtered, features engineered, and categorical data encoded."
        },
        {
            "item": "No data leakage between train and test sets",
            "status": "Yes",
            "notes": "The data was split before evaluation, and separate training and testing sets were used."
        },
        {
            "item": "Multiple modelling techniques were compared",
            "status": "Yes",
            "notes": "Logistic Regression and Random Forest were compared, and Random Forest was additionally tuned."
        },
        {
            "item": "The evaluation strategy was appropriate",
            "status": "Yes",
            "notes": "An 80/20 split and 5-fold cross-validation were used."
        },
        {
            "item": "The results support the business objective",
            "status": "Partial",
            "notes": "The model supports pattern analysis but is not yet strong enough for high-stakes operational deployment."
        }
    ],
    "overall_quality": "Good",
    "areas_for_improvement": [
        "Incorporate additional external features",
        "Address class imbalance more explicitly",
        "Explore more advanced models"
    ]
}

In [13]:
# Display the process review
print("=== Process Review Checklist ===")
for item in process_review["checklist"]:
    print(f"[{item['status']:^8}] {item['item']}")
    if item["notes"]:
        print(f"Note: {item['notes']}")
    print()

print(f"Overall Quality: {process_review['overall_quality']}")

if process_review["areas_for_improvement"]:
    print("\nAreas for Improvement:")
    for area in process_review["areas_for_improvement"]:
        print(f"- {area}")

=== Process Review Checklist ===
[Partial ] All relevant data sources were considered
Note: The project used a single crime dataset. Additional external data such as socio-economic or weather data could further improve performance.

[  Yes   ] Data preparation was thorough and documented
Note: Missing values were handled, duplicates removed, outliers filtered, features engineered, and categorical data encoded.

[  Yes   ] No data leakage between train and test sets
Note: The data was split before evaluation, and separate training and testing sets were used.

[  Yes   ] Multiple modelling techniques were compared
Note: Logistic Regression and Random Forest were compared, and Random Forest was additionally tuned.

[  Yes   ] The evaluation strategy was appropriate
Note: An 80/20 split and 5-fold cross-validation were used.

[Partial ] The results support the business objective
Note: The model supports pattern analysis but is not yet strong enough for high-stakes operational deployment.



The overall CRISP-DM process was followed successfully, and the project demonstrates a clear workflow from business understanding to modelling and evaluation. Data preparation was thorough, model comparison was appropriate, and no major evidence of data leakage was identified. However, the process could be improved further by integrating additional data sources and applying stronger techniques for handling class imbalance.

---
### Task 3: Determine Next Steps

Based on the evaluation results and process review, decide on the next course of action. The three main options are:

1. **Proceed to Deployment:** The model meets all criteria and is ready for deployment.
2. **Iterate:** Go back to a previous phase (e.g., Data Preparation or Modelling) to improve the model before deployment.
3. **Terminate:** The project is not viable in its current form and should be re-scoped or abandoned.

**Instructions:** Document your decision and provide a clear rationale.

In [14]:
# TODO: Determine and document the next steps.

next_steps = {
    "decision": "Iterate",
    "rationale": "The model demonstrates stable performance and useful analytical value, but the predictive accuracy is still too modest for deployment in a critical decision-making context.",
    "if_iterating": {
        "return_to_phase": "Phase 3: Data Preparation and Phase 4: Modeling",
        "specific_actions": [
            "Apply class balancing techniques such as SMOTE or class weighting",
            "Add external features such as socio-economic or environmental data",
            "Explore more advanced models such as XGBoost or Gradient Boosting",
            "Perform additional feature engineering and selection"
        ]
    },
    "if_deploying": {
        "selected_model": "Random Forest",
        "deployment_readiness": "Not ready for full operational deployment"
    }
}

In [15]:
# Display the decision
print("=" * 60)
print("NEXT STEPS DECISION")
print("=" * 60)

print(f"\nDecision  : {next_steps['decision']}")
print(f"Rationale : {next_steps['rationale']}")

if next_steps["decision"] == "Iterate":
    print(f"\nReturn to: {next_steps['if_iterating']['return_to_phase']}")
    print("Specific Actions:")
    for action in next_steps["if_iterating"]["specific_actions"]:
        print(f"- {action}")

elif next_steps["decision"] == "Proceed to Deployment":
    print(f"\nModel to deploy: {next_steps['if_deploying']['selected_model']}")
    print(f"Readiness      : {next_steps['if_deploying']['deployment_readiness']}")

NEXT STEPS DECISION

Decision  : Iterate
Rationale : The model demonstrates stable performance and useful analytical value, but the predictive accuracy is still too modest for deployment in a critical decision-making context.

Return to: Phase 3: Data Preparation and Phase 4: Modeling
Specific Actions:
- Apply class balancing techniques such as SMOTE or class weighting
- Add external features such as socio-economic or environmental data
- Explore more advanced models such as XGBoost or Gradient Boosting
- Perform additional feature engineering and selection


The recommended next step is to iterate rather than deploy immediately. Although the Random Forest model is stable and performs better than the baseline, the overall weighted F1-score remains modest due to class imbalance and the complexity of multi-class crime prediction. Returning to data preparation and modelling will allow the project to incorporate stronger features and improved modelling strategies.

Overall, the evaluation phase shows that the project successfully followed the CRISP-DM methodology and produced a stable predictive model for crime classification. The model provides useful insights into crime patterns and demonstrates reasonable generalisation ability through cross-validation. However, the relatively low performance metrics indicate that the problem remains challenging and that further iterations are needed before deployment. This evaluation therefore supports continued refinement of both the data preparation and modelling stages.